src/analysis.ts

```diff
@@ -61,3 +109,138 @@ export class Analysis {
     }
   }
 }
+
+
+class PackageGatherer implements AnalysisVisitor {
+  packageDirs = new Set<string>();
+  visitDocumentDescriptor(dd: DocumentDescriptor): void {
+    if (dd.document instanceof JsonDocument &&
+        (dd.document.url.endsWith('package.json') ||
+         dd.document.url.endsWith('bower.json'))) {
+      const dirname = path.dirname(dd.document.url);
+      if (!this.packageDirs.has(dirname)) {
+        this.packageDirs.add(trimLeft(dirname, '/'));
+      }
+    }
+  }
+}
+
+class ElementGatherer implements AnalysisVisitor {
+  elementDescriptors: ElementDescriptor[] = [];
+  locationOffsets = new Map<ElementDescriptor, LocationOffset>();
+  elementPaths = new Map<ElementDescriptor, string>();
+  visitElement(elementDescriptor: ElementDescriptor, path: Descriptor[]): void {
+    let pathToElement: string|null = null;
+    let locationOffset: LocationOffset|undefined;
+    for (const descriptor of path) {
+      if (descriptor instanceof DocumentDescriptor) {
+        pathToElement = descriptor.document.url;
+        locationOffset = descriptor.locationOffset;
+      }
+    }
+    if (!pathToElement) {
+      throw new Error(
+          `Unable to determine path to element: ${elementDescriptor}`);
+    }
+    if (this.elementPaths.has(elementDescriptor)) {
+      if (this.elementPaths.get(elementDescriptor) !== pathToElement) {
+        throw new Error(
+            `Found element ${elementDescriptor} at distinct paths: ` +
+            `${pathToElement} and ${this.elementPaths.get(elementDescriptor)}`);
+      }
+      return;
+    }
+
+    this.elementPaths.set(elementDescriptor, pathToElement);
+    this.locationOffsets.set(elementDescriptor, locationOffset);
+    this.elementDescriptors.push(elementDescriptor);
+  }
+}
+
+abstract class AnalysisVisitor {
+  visitDocumentDescriptor?(dd: DocumentDescriptor, path: Descriptor[]): void;
+  visitInlineDocumentDescriptor?
+      (dd: InlineDocumentDescriptor<any>, path: Descriptor[]): void;
+  visitElement?(element: ElementDescriptor, path: Descriptor[]): void;
+  visitImportDescriptor?
+      (importDesc: ImportDescriptor<any>, path: Descriptor[]): void;
+  done?(): void;
+}
+
+class AnalysisWalker {
```

Comment:
comment? Based on name alone I wonder what this difference is between this and Analysis Visitor